spring-rabbit-test/src/main/java/org/springframework/amqp/rabbit/test/RabbitListenerTestHarness.java

```diff
@@ -0,0 +1,176 @@
+import org.springframework.core.annotation.AnnotationAttributes;
+import org.springframework.core.annotation.AnnotationUtils;
+import org.springframework.core.type.AnnotationMetadata;
+import org.springframework.util.Assert;
+import org.springframework.util.StringUtils;
+
+/**
+ * BeanPostProcessor extending {@link RabbitListenerAnnotationBeanPostProcessor}.
+ * Wraps the listener bean in a CGLIB proxy with an advice to capture the arguments
+ * and result (if any) in a blocking queue. Test cases can access the results
+ * by autowiring the test harness into test cases.
+ *
+ * @author Gary Russell
+ * @since 1.6
+ *
+ */
+public class RabbitListenerTestHarness extends RabbitListenerAnnotationBeanPostProcessor {
+
+	private final Log logger = LogFactory.getLog(this.getClass());
+
+	private final Map<String, CaptureAdvice> listenerCapture = new HashMap<String, CaptureAdvice>();
+
+	private final AnnotationAttributes attributes;
+
+	private final Map<String, Object> listeners = new HashMap<String, Object>();
+
+	public RabbitListenerTestHarness(AnnotationMetadata importMetadata) {
+		Map<String, Object> map = importMetadata.getAnnotationAttributes(RabbitListenerTest.class.getName());
+		this.attributes = AnnotationAttributes.fromMap(map);
+		Assert.notNull(this.attributes,
+				"@RabbitListenerTest is not present on importing class " + importMetadata.getClassName());
+	}
+
+	@Override
+	protected void processListener(MethodRabbitListenerEndpoint endpoint, RabbitListener rabbitListener, Object bean,
+			Object adminTarget, String beanName) {
+		String id = rabbitListener.id();
+		if (StringUtils.hasText(id)) {
+			if (this.attributes.getBoolean("spy")) {
+				Object spy = Mockito.spy(bean);
+				this.listeners.put(id, spy);
+				super.processListener(endpoint, rabbitListener, spy, adminTarget, beanName);
+				return;
+			}
+			else {
+				try {
+					ProxyFactoryBean pfb = new ProxyFactoryBean();
+					pfb.setProxyTargetClass(true);
+					pfb.setTarget(bean);
+					CaptureAdvice advice = new CaptureAdvice();
+					pfb.addAdvice(advice);
+					Object advised = pfb.getObject();
+						this.listenerCapture.put(id, advice);
+						super.processListener(endpoint, rabbitListener, advised, adminTarget, beanName);
+						return;
+				}
+				catch (Exception e) {
+					logger.error("Failed to proxy @RabbitListener with id: " + id);
+				}
+			}
+		}
+		else {
+			logger.info("The test harness can only proxy @RabbitListeners with an 'id' attribute");
+		}
+		super.processListener(endpoint, rabbitListener, bean, adminTarget, beanName);
+	}
+
+	public InvocationData getNextInvocationDataFor(String id, long wait, TimeUnit unit) throws InterruptedException {
+		CaptureAdvice advice = this.listenerCapture.get(id);
+		if (advice != null) {
+			return advice.invocationData.poll(wait, unit);
+		}
+		return null;
+	}
+
+	@SuppressWarnings("unchecked")
+	<T> T getSpy(String id) {
+		return (T) this.listeners.get(id);
+	}
```

Comment:
How about to override `postProcessAfterInitialization()` and return `spy/proxy` directly instead of `bean`?

With that we still get a gain of `@Autowired MyListener listener;` and no need to invoke such a method from test.

As well as we could register `CaptureAdvice` (or anything else) as a bean for the listener.

I don't mind the current solution, just some brainstorm ideas.